src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,528 @@
+        }
+
+        /// <summary>
+        ///   Gets the closest contacts to the provided contact.
+        /// </summary>
+        /// <param name="id">
+        ///   The unique <see cref="IContact.Id"/> of a contact.
+        /// </param>
+        /// <returns>
+        ///   An ordered sequence of contact, sorted by closeness. 
+        /// </returns>
+        /// <remarks>
+        ///   "Closest" is the XOR metric of the contact.
+        /// </remarks>
+        public IEnumerable<T> Closest(byte[] id)
+        {
+            return this.Select(c => new
+                {
+                    distance = Distance(c.Id, id), contact = c
+                })
+               .OrderBy(a => a.distance).Select(a => a.contact);
+        }
+
+        /// <inheritdoc />
+        public int Count => Root.DeepCount();
+
+        /// <inheritdoc />
+        public bool IsReadOnly => false;
+
+        /// <inheritdoc />
+        public void Add(T item)
+        {
+            Validate(item);
+            bool q;
+            PingEventArgs<T> e;
+            _rwlock.EnterWriteLock();
+            try
+            {
+                q = _Add(item, out e);
+            }
+            finally
+            {
+                _rwlock.ExitWriteLock();
+            }
+
+            // Could not add.  Ping oldest contacts.
+            if (!q) Ping?.Invoke(this, e);
+        }
+
+        /// <inheritdoc />
+        public void Clear() { Root = new Bucket<T>(); }
+
+        /// <inheritdoc />
+        public bool Contains(T item)
+        {
+            Validate(item);
+
+            _rwlock.EnterReadLock();
+            try
+            {
+                return _Get(item.Id) != null;
+            }
+            finally
+            {
+                _rwlock.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        ///   Gets the contact associated with the specified ID.
+        /// </summary>
+        /// <param name="id">
+        ///   The ID of an <see cref="IContact"/>.
+        /// </param>
+        /// <param name="contact">
+        ///   When this method returns, contains the <see cref="IContact"/> associated 
+        ///   with the <paramref name="id"/>, if the key is found; otherwise, <b>null</b>.
+        ///   This parameter is passed uninitialized.
+        /// </param>
+        /// <returns>
+        ///  <b>true</b> if the <paramref name="id"/> is found; otherwise <b>false</b>.
+        /// </returns>
+        public bool TryGet(byte[] id, out T contact)
+        {
+            _rwlock.EnterReadLock();
+            try
+            {
+                contact = _Get(id);
+                return contact != null;
+            }
+            finally
+            {
+                _rwlock.ExitReadLock();
+            }
+        }
+
+        /// <inheritdoc />
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            foreach (var contact in this) array[arrayIndex++] = contact;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'foreach' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)